### ISSUE TITLE
Too many open files error  while operating (ls command) on  a large collection of datasets

### ISSUE THREAD

**Issue by yarikoptic** [datalad/datalad#1888]
#### What is the problem?
`datalad ls -r ` on a sizeable hierarchy eventually just starts spitting out lots of errors
```
...
inbox/2016/12/05/unknown           [annex]  master  - 2017-10-05/18:43:25  OK
inbox/2016/12/06                   [git]  master  - 2017-10-05/19:33:48  OK
inbox/2016/12/06/A000124           Cmd('/home/yoh/git-annex-standalone/usr/lib/git-annex.linux/git') not found due to: OSError('[Errno 24] Too many open files')
  cmdline: /home/yoh/git-annex-standalone/usr/lib/git-annex.linux/git -c receive.autogc=0 -c gc.auto=0 cat-file --batch [cmd.py:execute:735]
[ERROR] Failed to start ['git', '-c', 'receive.autogc=0', '-c', 'gc.auto=0', 'annex', 'status', '--json']'': [Errno 24] Too many open files [subprocess.py:pipe_cloexec:860]
...
```

I wonder if we are doomed to set some kind of an upper bound on how many repos/datasets we keep in the flyweight cache, and prune some least recently referenced?  or how to overcome (tried unassigning .repo as soon as used up etc, didn't help)?

**Comment by yarikoptic** [comment]
so the cause is a bunch of `git cat-file --batch-check` and `cat-file --batch` processes started by something, I don't know yet what (GitPython or git-annex) -- there is no git-annex processes parent to them....

doing .repo.repo.close() seems to do the trick!  will add it as a workaround in ls for now

**Comment by bpoldrack** [comment]
Re workaround:
That's what `*Repo.repo.git.clear_cache()` is for. ;-)

**Comment by mih** [comment]
Fixed in 85b5c069c0affafc71035576bcd0109e92e80bc9
